File: desktop/electron/backend/orchestrator/teachingMemory.ts
/**
 * Long-horizon teaching memory.
 *
 * Adapted for Socratic teaching from the "handoff summary" pattern used by
 * OpenAI Codex' local compaction path, and the "model-visible means logged"
 * principle of DeepSeek Harness:
 *
 *  - Recent turns are kept verbatim (bounded window, oldest folds out).
 *  - When the verbatim window reaches its budget, turns are folded into a
 *    rolling structured handoff summary written in "notes-to-self" register so
 *    a successor request resumes naturally instead of restarting (Codex
 *    "handoff > summary": forward-looking — progress / constraints / next steps).
 *  - Compaction can be LLM-driven via {@link TeachingMemory.compactWithLlm},
 *    matching Codex' local compaction path, or fall back to naive truncation.
 *  - Raw history is never destroyed: every turn is kept in an append-only log
 *    (Harness "model-visible means logged"); pruning only changes the surface
 *    the model sees, so audit/replay/persistence stay possible.
 *  - The emitted context block enforces a soft token budget (Codex
 *    `_enforce_token_budget`), keeping the newest, most task-relevant content.
 *
 * The model-visible block emitted by {@link TeachingMemory.toContextBlock} is
 * injected into both the routing LLM input and the Collaboration agent so the
 * teacher can recall what was already covered across a long session.
 */

/** A single completed teaching exchange. */
export interface TeachingTurn {
  /** What the student said (verbatim). */
  student: string
  /** What the teacher replied (the delivered response). */
  teacher: string
  /** Pedagogical action taken this turn. */
  action: string
  /** Hint level after this turn. */
  hintLevel: number
  /** Concept currently being worked on, if known. */
  concept?: string
}

export interface TeachingMemoryOptions {
  /** Max number of turns kept verbatim in the window. */
  maxVerbatimTurns?: number
  /** Max length (chars) of the folded rolling summary. */
  maxSummaryLength?: number
  /** Soft token budget for the whole session; negative disables the cap. */
  sessionTokenBudget?: number
  /** Soft token budget for the emitted model-visible block; <=0 disables. */
  maxBlockTokens?: number
}

/**
 * Handoff compaction prompt (adapted from Codex' context-checkpoint prompt).
 * Written to produce a forward-looking handoff, not a backward-looking recap.
 */
export const HANDOFF_PROMPT = `You are performing a CONTEXT CHECKPOINT for a long-running Socratic math teaching session.
Create a handoff summary for another LLM (the teacher) that will resume guiding this student.

Include:
- 已达成进度 (progress): what the student has accomplished and understood so far
- 关键决策与约束 (constraints): pedagogical decisions, difficulty calibration, pacing rules
- 学生状态 (student state): mastery level, common mistakes, emotional/engagement signals, learning habits
- 下一步 (next steps): what to teach next, which concepts remain uncovered

Write in concise notes-to-self register (Chinese), not exposition. Preserve any concrete facts (numbers, exact student phrasing) that the successor needs to continue accurately.`

/**
 * Rough token estimate used for block-budget enforcement. CJK-heavy text runs
 * ~1 token per character; latin ~4 chars/token. Dividing by 4 is a safe,
 * over-estimating heuristic that keeps the block under budget.
 */
export function estimateTokens(text: string): number {
  if (!text) return 0
  return Math.ceil(text.length / 4)
}

export class TeachingMemory {
  readonly maxVerbatimTurns: number
  readonly maxSummaryLength: number
  readonly sessionTokenBudget: number
  readonly maxBlockTokens: number

  private recentTurns: TeachingTurn[] = []
  /** Append-only log of every turn (Harness: raw history is never destroyed). */
  private allTurns: TeachingTurn[] = []
  private rollingSummary = ''
  private coveredConcepts: string[] = []
  private hintLevel = 0
  private totalTokensUsed = 0

  constructor(opts: TeachingMemoryOptions = {}) {
    this.maxVerbatimTurns = opts.maxVerbatimTurns ?? 6
    this.maxSummaryLength = opts.maxSummaryLength ?? 1200
    this.sessionTokenBudget = opts.sessionTokenBudget ?? -1
    this.maxBlockTokens = opts.maxBlockTokens ?? -1
  }

  /** Acknowledged concepts, deduplicated, in first-seen order. */
  get concepts(): string[] {
    return [...new Set(this.coveredConcepts)]
  }

  /** Current hint level (latest turn). */
  get currentHintLevel(): number {
    return this.hintLevel
  }

  /** Cumulative tokens consumed by LLM calls this session. */
  get tokensUsed(): number {
    return this.totalTokensUsed
  }

  /** Whether the session has exceeded its configured soft token budget. */
  get overBudget(): boolean {
    return this.sessionTokenBudget >= 0 && this.totalTokensUsed > this.sessionTokenBudget
  }

  /** Number of turns currently held verbatim. */
  get verbatimTurnCount(): number {
    return this.recentTurns.length
  }

  /** Read-only copy of the append-only log (all turns, never pruned). */
  get log(): ReadonlyArray<TeachingTurn> {
    return this.allTurns
  }

  /** Number of turns ever recorded (== log length). */
  get totalTurnCount(): number {
    return this.allTurns.length
  }

  /**
   * Preventive compaction trigger (Codex: compact before the window fills,
   * not reactively after overflow). True once the verbatim window is at or
   * beyond its budget, meaning the next turn would start folding.
   */
  shouldCompact(): boolean {
    return this.recentTurns.length >= this.maxVerbatimTurns
  }

  /** Record one completed teaching turn and fold/trim as needed. */
  recordTurn(turn: TeachingTurn, usageTokens = 0): void {
    this.totalTokensUsed += usageTokens
    this.allTurns.push(turn)
    this.recentTurns.push(turn)
    if (turn.concept) {
      this.coveredConcepts.push(turn.concept)
    }
    this.hintLevel = turn.hintLevel
    this.prune()
  }

  /**
   * LLM-driven handoff compaction (Codex local compaction path): fold the whole
   * verbatim window into a structured handoff summary via the provided
   * summarizer, so a successor request resumes naturally. Falls back to naive
   * truncation when the summarizer returns nothing or throws.
   *
   * @returns the folded rolling summary.
   */
  async compactWithLlm(
    summarizer: (prompt: string, turns: ReadonlyArray<TeachingTurn>) => Promise<string> | string,
  ): Promise<string> {
    if (this.recentTurns.length === 0) return this.rollingSummary
    const turns = this.recentTurns.splice(0, this.recentTurns.length)
    let folded = ''
    try {
      const summary = await summarizer(HANDOFF_PROMPT, turns)
      folded = typeof summary === 'string' ? summary.trim() : ''
    } catch {
      folded = ''
    }
    if (folded) {
      this.rollingSummary = this.rollingSummary
        ? `${this.rollingSummary}\n---\n${folded}`
        : folded
    } else {
      // Fallback: naive fold of the evicted turns.
      for (const t of turns) {
        const line = `学生: ${t.student}\n教师: ${t.teacher}（${t.action}）`
        this.rollingSummary = this.rollingSummary ? `${this.rollingSummary}\n---\n${line}` : line
      }
    }
    this.trimSummary()
    return this.rollingSummary
  }

  /** Fold the oldest turns into the rolling summary until the window fits. */
  private prune(): void {
    while (this.recentTurns.length > this.maxVerbatimTurns) {
      const oldest = this.recentTurns.shift()!
      const line = `学生: ${oldest.student}\n教师: ${oldest.teacher}（${oldest.action}）`
      this.rollingSummary = this.rollingSummary
        ? `${this.rollingSummary}\n---\n${line}`
        : line
    }
    this.trimSummary()
  }

  /** Keep the summary within a bounded window (newest tail). */
  private trimSummary(): void {
    if (this.rollingSummary.length > this.maxSummaryLength) {
      this.rollingSummary = this.rollingSummary.slice(-this.maxSummaryLength)
    }
  }

  /** Reset the memory (used when a new session starts). */
  reset(): void {
    this.recentTurns = []
    this.allTurns = []
    this.rollingSummary = ''
    this.coveredConcepts = []
    this.hintLevel = 0
    this.totalTokensUsed = 0
  }

  /**
   * Model-visible context block. Injected into LLM prompts so both the routing
   * decision and the teacher's synthesis can recall prior turns. When
   * `maxTokens > 0`, the block is trimmed to that soft budget keeping the
   * newest, most task-relevant content (concepts + hint level always survive).
   */
  toContextBlock(maxTokens = this.maxBlockTokens): string {
    const head: string[] = []
    const concepts = this.concepts
    if (concepts.length > 0) head.push(`已覆盖概念: ${concepts.join(', ')}`)
    head.push(`当前提示等级: ${this.hintLevel}`)

    let summary = this.rollingSummary
    let recent = [...this.recentTurns]

    const render = (): string => {
      const parts = [...head]
      if (summary) parts.push(`较早对话摘要:\n${summary}`)
      if (recent.length > 0) {
        parts.push('近几轮对话:')
        for (const t of recent) parts.push(`- 学生: ${t.student}\n  教师: ${t.teacher}`)
      }
      return parts.join('\n')
    }

    if (maxTokens > 0) {
      // 1) Shrink the folded summary (newest tail kept) until it fits.
      while (summary && estimateTokens(render()) > maxTokens) {
        summary = summary.slice(Math.max(0, Math.floor(summary.length * 0.6)))
      }
      // 2) Drop the oldest verbatim turns until it fits.
      while (recent.length > 0 && estimateTokens(render()) > maxTokens) {
        recent.shift()
      }
    }
    return render()
  }

  /**
   * 序列化为可持久化的普通对象（跨会话恢复）。
   *
   * 输出包含追加式轮次日志 allTurns 与滚动摘要，写入 StateStore 后可在新的
   * 会话中重建，实现长周期教学任务的续接（对应 Harness "模型可见即已记录"
   * —— 原始历史不因压缩而丢失）。
   */
  toJSON(): TeachingMemorySnapshot {
    return {
      recentTurns: this.recentTurns,
      allTurns: this.allTurns,
      rollingSummary: this.rollingSummary,
      coveredConcepts: this.coveredConcepts,
      hintLevel: this.hintLevel,
      totalTokensUsed: this.totalTokensUsed,
    }
  }

  /**
   * 从持久化快照重建教学记忆（配合 {@link toJSON}）。
   *
   * 对缺失/畸形字段做防御性兜底，即使旧版本数据也能安全恢复，不会抛错。
   */
  static fromJSON(data: Partial<TeachingMemorySnapshot>, opts: TeachingMemoryOptions = {}): TeachingMemory {
    const m = new TeachingMemory(opts)
    m.recentTurns = Array.isArray(data.recentTurns) ? data.recentTurns : []
    m.allTurns = Array.isArray(data.allTurns) ? data.allTurns : []
    m.rollingSummary = typeof data.rollingSummary === 'string' ? data.rollingSummary : ''
    m.coveredConcepts = Array.isArray(data.coveredConcepts) ? data.coveredConcepts : []
    m.hintLevel = typeof data.hintLevel === 'number' ? data.hintLevel : 0
    m.totalTokensUsed = typeof data.totalTokensUsed === 'number' ? data.totalTokensUsed : 0
    return m
  }
}

/** TeachingMemory 的可序列化快照（用于跨会话持久化）。 */
export interface TeachingMemorySnapshot {
  recentTurns: TeachingTurn[]
  allTurns: TeachingTurn[]
  rollingSummary: string
  coveredConcepts: string[]
  hintLevel: number
  totalTokensUsed: number
}

File: backend/mathweaver/orchestrator/teaching_memory.py
"""Long-horizon teaching memory.

Adapted for Socratic teaching from the "handoff summary" pattern used by
OpenAI Codex' local compaction path, and the "model-visible means logged"
principle of DeepSeek Harness:

- Recent turns are kept verbatim (bounded window, oldest folds out).
- When the verbatim window reaches its budget, turns are folded into a rolling
  structured handoff summary written in "notes-to-self" register so a successor
  request resumes naturally instead of restarting (Codex "handoff > summary":
  forward-looking --- progress / constraints / next steps).
- Compaction can be LLM-driven via ``compact_with_llm`` (Codex local compaction
  path), or fall back to naive truncation.
- Raw history is never destroyed: every turn is kept in an append-only log
  (Harness "model-visible means logged"); pruning only changes the surface the
  model sees, so audit/replay/persistence stay possible.
- The emitted context block enforces a soft token budget (Codex
  ``_enforce_token_budget``), keeping the newest, most task-relevant content.

The model-visible block emitted by ``to_context_block`` is injected into both
the routing LLM input and the Collaboration agent so the teacher can recall
what was already covered across a long session.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable, Sequence
from dataclasses import dataclass

#: Handoff compaction prompt (adapted from Codex' context-checkpoint prompt).
#: Written to produce a forward-looking handoff, not a backward-looking recap.
HANDOFF_PROMPT = """You are performing a CONTEXT CHECKPOINT for a long-running Socratic math teaching session.
Create a handoff summary for another LLM (the teacher) that will resume guiding this student.

Include:
- 已达成进度 (progress): what the student has accomplished and understood so far
- 关键决策与约束 (constraints): pedagogical decisions, difficulty calibration, pacing rules
- 学生状态 (student state): mastery level, common mistakes, emotional/engagement signals, learning habits
- 下一步 (next steps): what to teach next, which concepts remain uncovered

Write in concise notes-to-self register (Chinese), not exposition. Preserve any concrete facts (numbers, exact student phrasing) that the successor needs to continue accurately."""


@dataclass
class TeachingTurn:
    """A single completed teaching exchange."""

    #: What the student said (verbatim).
    student: str
    #: What the teacher replied (the delivered response).
    teacher: str
    #: Pedagogical action taken this turn.
    action: str
    #: Hint level after this turn.
    hint_level: int
    #: Concept currently being worked on, if known.
    concept: str | None = None


def estimate_tokens(text: str) -> int:
    """Rough token estimate used for block-budget enforcement.

    CJK-heavy text runs ~1 token per character; latin ~4 chars/token. Dividing
    by 4 is a safe, over-estimating heuristic that keeps the block under budget.
    """
    if not text:
        return 0
    return (len(text) + 3) // 4


class TeachingMemory:
    """Per-session rolling memory of the teaching conversation."""

    def __init__(
        self,
        max_verbatim_turns: int = 6,
        max_summary_length: int = 1200,
        session_token_budget: int = -1,
        max_block_tokens: int = -1,
    ) -> None:
        self.max_verbatim_turns = max_verbatim_turns
        self.max_summary_length = max_summary_length
        self.session_token_budget = session_token_budget
        self.max_block_tokens = max_block_tokens
        self._recent_turns: list[TeachingTurn] = []
        #: Append-only log of every turn (Harness: raw history is never destroyed).
        self._all_turns: list[TeachingTurn] = []
        self._rolling_summary = ""
        self._covered_concepts: list[str] = []
        self._hint_level = 0
        self._total_tokens_used = 0

    # -- Read-only accessors ------------------------------------------------

    @property
    def concepts(self) -> list[str]:
        """Acknowledged concepts, deduplicated, in first-seen order."""
        return list(dict.fromkeys(self._covered_concepts))

    @property
    def current_hint_level(self) -> int:
        return self._hint_level

    @property
    def tokens_used(self) -> int:
        return self._total_tokens_used

    @property
    def over_budget(self) -> bool:
        return self.session_token_budget >= 0 and self._total_tokens_used > self.session_token_budget

    @property
    def verbatim_turn_count(self) -> int:
        return len(self._recent_turns)

    @property
    def log(self) -> list[TeachingTurn]:
        """Read-only copy of the append-only log (all turns, never pruned)."""
        return list(self._all_turns)

    @property
    def total_turn_count(self) -> int:
        """Number of turns ever recorded (== log length)."""
        return len(self._all_turns)

    # -- Mutation -----------------------------------------------------------

    def should_compact(self) -> bool:
        """Preventive compaction trigger (Codex: compact before the window fills).

        True once the verbatim window is at or beyond its budget, meaning the
        next turn would start folding.
        """
        return len(self._recent_turns) >= self.max_verbatim_turns

    def record_turn(self, turn: TeachingTurn, usage_tokens: int = 0) -> None:
        """Record one completed teaching turn and fold/trim as needed."""
        self._total_tokens_used += usage_tokens
        self._all_turns.append(turn)
        self._recent_turns.append(turn)
        if turn.concept:
            self._covered_concepts.append(turn.concept)
        self._hint_level = turn.hint_level
        self._prune()

    async def compact_with_llm(
        self,
        summarizer: Callable[[str, Sequence[TeachingTurn]], str | Awaitable[str]],
    ) -> str:
        """LLM-driven handoff compaction (Codex local compaction path).

        Fold the whole verbatim window into a structured handoff summary via the
        provided summarizer so a successor request resumes naturally. Falls back
        to naive truncation when the summarizer returns nothing or raises.

        Returns the folded rolling summary.
        """
        if not self._recent_turns:
            return self._rolling_summary
        turns = self._recent_turns
        self._recent_turns = []
        folded = ""
        try:
            result = summarizer(HANDOFF_PROMPT, turns)
            if isinstance(result, Awaitable):
                result = await result
            folded = result.strip() if isinstance(result, str) else ""
        except Exception:
            folded = ""
        if folded:
            self._rolling_summary = (
                f"{self._rolling_summary}\n---\n{folded}" if self._rolling_summary else folded
            )
        else:
            # Fallback: naive fold of the evicted turns.
            for t in turns:
                line = f"学生: {t.student}\n教师: {t.teacher}（{t.action}）"
                self._rolling_summary = (
                    f"{self._rolling_summary}\n---\n{line}" if self._rolling_summary else line
                )
        self._trim_summary()
        return self._rolling_summary

    def _prune(self) -> None:
        """Fold the oldest turns into the rolling summary until the window fits."""
        while len(self._recent_turns) > self.max_verbatim_turns:
            oldest = self._recent_turns.pop(0)
            line = f"学生: {oldest.student}\n教师: {oldest.teacher}（{oldest.action}）"
            self._rolling_summary = (
                f"{self._rolling_summary}\n---\n{line}" if self._rolling_summary else line
            )
        self._trim_summary()

    def _trim_summary(self) -> None:
        """Keep the summary within a bounded window (newest tail)."""
        if len(self._rolling_summary) > self.max_summary_length:
            self._rolling_summary = self._rolling_summary[-self.max_summary_length :]

    def reset(self) -> None:
        """Reset the memory (used when a new session starts)."""
        self._recent_turns = []
        self._all_turns = []
        self._rolling_summary = ""
        self._covered_concepts = []
        self._hint_level = 0
        self._total_tokens_used = 0

    # -- Rendering ----------------------------------------------------------

    def to_context_block(self, max_tokens: int | None = None) -> str:
        """Model-visible context block. Injected into LLM prompts.

        When ``max_tokens`` resolves to a positive budget, the block is trimmed
        to that soft budget keeping the newest, most task-relevant content
        (concepts + hint level always survive).
        """
        if max_tokens is None:
            max_tokens = self.max_block_tokens
        head: list[str] = []
        concepts = self.concepts
        if concepts:
            head.append(f"已覆盖概念: {', '.join(concepts)}")
        head.append(f"当前提示等级: {self._hint_level}")

        summary = self._rolling_summary
        recent = list(self._recent_turns)

        def render() -> str:
            parts: list[str] = list(head)
            if summary:
                parts.append(f"较早对话摘要:\n{summary}")
            if recent:
                parts.append("近几轮对话:")
                for t in recent:
                    parts.append(f"- 学生: {t.student}\n  教师: {t.teacher}")
            return "\n".join(parts)

        if max_tokens and max_tokens > 0:
            # 1) Shrink the folded summary (newest tail kept) until it fits.
            while summary and estimate_tokens(render()) > max_tokens:
                summary = summary[max(len(summary) * 6 // 10, 0) :]
            # 2) Drop the oldest verbatim turns until it fits.
            while recent and estimate_tokens(render()) > max_tokens:
                recent.pop(0)
        return render()